### DISCUSSION TITLE
Add validation for dropdown values

### DISCUSSION THREAD

**Discussion by dpihac** [directus/directus#11802] [Feature Requests]
I have a collection that has a field of the following definition. There are three possible choices in the UI:

- EUR
- HRK
- USD

However, when using the REST API, any value can be defined and successfully inserted in the DB. There is no validation whatsoever.

```{
	"data": {
		"collection": "fondovi",
		"field": "valuta",
		"type": "string",
		"meta": {
			"id": 16,
			"collection": "fondovi",
			"field": "valuta",
			"special": null,
			"interface": "select-dropdown",
			"options": {
				"choices": [
					{
						"text": "EUR",
						"value": "EUR"
					},
					{
						"text": "HRK",
						"value": "HRK"
					},
					{
						"text": "USD",
						"value": "USD"
					}
				]
			},
			"display": null,
			"display_options": null,
			"readonly": false,
			"hidden": false,
			"sort": 3,
			"width": "half",
			"translations": null,
			"note": null,
			"conditions": null,
			"required": false,
			"group": "Information"
		},
		"schema": {
			"name": "valuta",
			"table": "fondovi",
			"schema": "public",
			"data_type": "character varying",
			"is_nullable": true,
			"generation_expression": null,
			"default_value": null,
			"is_generated": false,
			"max_length": 255,
			"comment": null,
			"numeric_precision": null,
			"numeric_scale": null,
			"is_unique": false,
			"is_primary_key": false,
			"has_auto_increment": false,
			"foreign_key_schema": null,
			"foreign_key_table": null,
			"foreign_key_column": null
		}
	}
}
```

The UI looks as follows:

![image](https://user-images.githubusercontent.com/2896514/155217110-1b529666-259e-4c9d-be3e-a610faaa0d87.png)

Notice that field "Valuta" is set to "NEW". When saving the item, there are no validation errors. UI correctly limits available values which can be selected:

![image](https://user-images.githubusercontent.com/2896514/155217606-220d09a6-c17e-4a01-94e9-2f46890569c0.png)

**Expected behavior:** prevent insert / update via the REST API if the value of the field is not in allowed values.

### To Reproduce

Following request:

```
POST /items/fondovi HTTP/1.1
Content-Type: application/json
Host: XXX
Content-Length: 104

{
	"status": "published",
	"naziv": "TEST FOR GITHUB",
	"drustvo": 1,
	"valuta": "THIS DOES NOT EXITS"
}
```

will create the item with the values as shown in the UI:

![image](https://user-images.githubusercontent.com/2896514/155218348-208933cc-0232-4d21-816d-880fa027cd09.png)

### What version of Directus are you using?

9.5.2

### What version of Node.js are you using?

16.13.1

### What database are you using?

PostgreSQL 12.9 on x86_64-pc-linux-gnu

### What browser are you using?

Firefox

### What operating system are you using?

Linux

### How are you deploying Directus?

Docker on Digital Ocean Apps

**Comment by azrikahar** [comment]
This is indeed the current behavior as the dropdown values are just "limited" in a sense from the UI perspective during user input, and it's definitely an enhancement to make this validation occur from the API side as well.

Although this is currently not supported yet, you should be able to utilize the [filter hook](https://docs.directus.io/extensions/hooks/#filter) to validate the input in the API. Here's an example:

### Collection/Field setup

- Collection: `products`
- Dropdown Field: `colors`
    - Choices:
        - red
        - green

![image](https://user-images.githubusercontent.com/42867097/155276915-1a466eaf-f38d-4f66-80d4-49d31215ce5f.png)

### Extension Code

(Make sure to build this extension as hooks does not support ESM syntax (`export default`) as of now)

```js
export default ({ filter }, { services, exceptions }) => {
	const { FieldsService } = services;
	const { InvalidPayloadException } = exceptions;

	filter('products.items.create', async (input, { collection }, { database, schema }) => {
		const productsFieldsService = new FieldsService({
			knex: database,
			schema,
		});
		
                // get the "colors" field in "products" collection and it's configured options
		const colorsField = await productsFieldsService.readOne(collection, 'colors');
		const options = colorsField.meta.options;
		
		// no need to do validation if allowOthers is true
		if (!options.allowOthers && input.colors) {
			const values = options.choices.map((choice) => choice.value);
			if (!values.includes(input.colors))
				throw new InvalidPayloadException(`colors must be one of (${values.join(', ')}).`);
		}

		return input;
	});
};
```

### POST request

We can see when we are sending colors that are _not_ `red` or `green`, it throws invalid payload error:

![chrome_8KEpNFShEu](https://user-images.githubusercontent.com/42867097/155277008-87ba542d-a89c-42c5-b465-302763ea3040.png)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
